cpp/include/rapidsmpf/streaming/core/message.hpp

```diff
@@ -0,0 +1,215 @@
+/**
+ * SPDX-FileCopyrightText: Copyright (c) 2025, NVIDIA CORPORATION & AFFILIATES.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#pragma once
+
+#include <any>
+#include <memory>
+#include <mutex>
+#include <stdexcept>
+#include <typeinfo>
+
+#include <rapidsmpf/error.hpp>
+
+namespace rapidsmpf::streaming {
+
+/**
+ * @brief Type-erased message wrapper around a payload.
+ *
+ * Holds a type-erased payload that may be shared with other messages. When the
+ * payload is shared, the handle still provides const access via `get<T>()`, but
+ * transferring ownership with `release<T>()` is only allowed if the message is
+ * solely owned.
+ *
+ * @note Copying is intentionally restricted; use `shallow_copy()` to create
+ * another handle that shares the same payload.
+ */
+class Message {
+    /**
+     * @brief Internal container for the (shared) payload.
+     */
+    struct Payload {
+        mutable std::mutex mutex;
+        /**
+         * @brief Type-erased holder for the actual payload.
+         *
+         * Stores a `std::shared_ptr<T>` for some `T`.
+         *
+         * @note Conceptually, a `std::unique_ptr` would suffice for exclusive ownership,
+         * but `std::any` requires its contents to be copyable, so a `std::shared_ptr` is
+         * used instead.
+         */
+        std::any data;
+    };
+
+  public:
+    /// @brief Create an empty message.
+    Message() = default;
+
+    /**
+     * @brief Construct a message from a unique pointer to a payload.
+     *
+     * Takes ownership of @p payload and stores it as a shared object internally.
+     *
+     * @tparam T Payload type.
+     * @param payload Non-null unique pointer to the payload instance.
+     * @throws std::invalid_argument if @p payload is null.
+     */
+    template <typename T>
+    Message(std::unique_ptr<T> payload) {
+        RAPIDSMPF_EXPECTS(
+            payload != nullptr, "payload cannot be null", std::invalid_argument
+        );
+        payload_ = std::make_shared<Payload>();
+        payload_->data = std::shared_ptr<T>(std::move(payload));
+    }
+
+    /** @brief Move construct. @param other Source message. */
+    Message(Message&& other) noexcept = default;
+
+    /** @brief Move assign. @param other Source message. @return *this. */
+    Message& operator=(Message&& other) noexcept = default;
+
+    /**
+     * @brief Reset the message to the empty state.
+     *
+     * Releases the stored payload (if this is the last shared owner).
+     *
+     * @note After this call, the message is empty.
+     */
+    void reset() noexcept {
+        return payload_.reset();
+    }
```

Comment:
Shouldn't we prevent concurrent calls to `reset` and `shallow_copy`?
```suggestion
    void reset() noexcept {
        {
        std::lock_guard lock(payload_->mtx); // maybe this lock should be an arg? 
        // first reset data while locking the mtx
        payload_->data.reset();
        }
        // now, any calls to Message will throw and we can safely reset the payload?
        payload_.reset();
    }
```

Reply:
reset only _resets_ this message **not** the payload. I don't think the message themselves should have an associated lock.

Update to docs:
```c++
    /**
     * @brief Resets the message to an empty state.
     *
     * If this is the last shared owner, the stored payload is deallocated.
     * Otherwise, the payload remains untouched and shared with other owners.
     *
     * @note After this call, the message becomes empty.
     */
    void reset() noexcept {
        return payload_.reset();
    }
```

Reply:
This is against the contract of the `shallow_copy` IMO, because we are supposed to throw if the payload is empty.

Reply:
> [@madsbk](https://github.com/madsbk) I still think this is problematic. For ex: say we have 1 thread calling `reset` and another calling `shallow_copy` and the ref count is 1 currently.
> 
> [[nodiscard]] Message shallow_copy() const {
>         auto lock = lock_payload();
>         return *this;
>     }
> 
> Say thread2 succeeds locking the pay load, but before the copy constructor, thread1 `reset`s payload. Since payload_ refcount is still 1, it'll deref the payload. Then, the copied message would be an empty with no payload.

I think I understand now, good catch. Yes, `reset()` needs to acquire the lock as well!